Compile error with Clang 13: std::result_of is removed in C++20

Forgive me if this has already been fixed in latest commits. But trying to build 4.5.3 with Clang 13 I get a compile error:
```
type_traits.hpp:87:35: error: no type named 'result_of' in namespace 'std'
using result_of_t = typename std::result_of<T>::type;
```
And so on. This is because the deprecated `std::result_of` has finally been removed in Clang 13. I think the fix would be to use `std::invoke_result` instead. (I do not have a patch but it would be great if someone could provide one as we are blocked in our effort to turn on C++20 by this issue)

Perhaps problem can be reproduced with OpenCV build:
- `cmake ...  -DCMAKE_CXX_STANDARD=20`
- build opencv_test_gapi

Note: it doesn't trigger `result_of` problem with GCC 10.3.1 or Clang 11.0.0 from Fedora 33

@dmatveev Could someone take a look on ADE build issues?

@daravi As workaround you can build OpenCV using C++14/17. This should not affect your applications which should be able to works with C++20 and use C++14/17 libraries ("ade" is internal dependency, it's headers are not exposed).

Another ADE build issues:

```
3rdparty/ade/ade-0.1.1f/sources/ade/include/ade/util/memory_range.hpp:237:24: warning: ‘template<class _Tp> struct std::is_pod’ is deprecated: use is_standard_layout && is_trivial instead [-Wdeprecated-declarations]
  237 |     static_assert(std::is_pod< util::decay_t<decltype(*data(src))> >::value, "Types must be pod");
      |                        ^~~~~~
```

--- a/modules/gapi/cmake/DownloadADE.cmake
+++ b/modules/gapi/cmake/DownloadADE.cmake
@@ -1,7 +1,7 @@
 set(ade_src_dir "${OpenCV_BINARY_DIR}/3rdparty/ade")
-set(ade_filename "v0.1.1f.zip")
-set(ade_subdir "ade-0.1.1f")
-set(ade_md5 "b624b995ec9c439cbc2e9e6ee940d3a2")
+set(ade_filename "v0.1.2.zip")
+set(ade_subdir "ade-0.1.2")
+set(ade_md5 "561c1e28ccf27ad0557a18e251c22226")
 ocv_download(FILENAME ${ade_filename}
              HASH ${ade_md5}
              URL

